mesonbuild/utils/universal.py

```diff
@@ -186,10 +186,23 @@ class _VerPickleLoadable(Protocol):
     python_command = [sys.executable]
 _meson_command: T.Optional['ImmutableListProtocol[str]'] = None
 
-if sys.version_info >= (3, 11):
-    getencoding = locale.getencoding
-else:
-    getencoding = locale.getpreferredencoding
+
+@lru_cache()
+def getencoding() -> str:
+    if sys.platform == 'win32':  # is_windows()
+        from ctypes import windll
+
+        try:
+            # This seems the only reliable way to get current console encoding
```

Comment:
Other than the fact that getpreferredencoding bad, and you spent a few commit messages adding something that apparently doesn't even work and then backing it out....what was wrong with the way we were doing this before this PR?

Reply:
Previous attempt worked, as it allowed the unittests to pass both on my machine and on CI. But it prevented using utf-8 when the console allowed it, since it used `cp1252` instead of `cp65001`, even if I called `chcp 65001`. Console encoding on Windows is a big mess, and I'm still not 100% sure what is the right approach...